### ISSUE TITLE
Simultaniously create limit on AWS dynamodb create

### ISSUE THREAD

**Issue by nenti** [serverless/serverless#3744]
Hi, AWS seems to have introduced a limit on creating dynamodb databases simultaniously. This needs a different handling of database create feature of serverless framework. 

```
Serverless Error ---------------------------------------

   An error occurred while provisioning your stack: ProductsTable
   - Subscriber limit exceeded: You have exceeded the maximum
   number of indexed tables that can be created simultaneously.
```

Locally I have solved this problem by putting a timeout inbetween each table reaction of 5 seconds.

--------------
http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html
```
CreateTable/UpdateTable/DeleteTable
In general, you can have up to 10 CreateTable, UpdateTable, and DeleteTable requests running simultaneously (in any combination). In other words, the total number of tables in the CREATING, UPDATING or DELETING state cannot exceed 10.
```

**Comment by HyperBrain** [comment]
The current limits as per http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html are:

> In general, you can have up to 10 CreateTable, UpdateTable, and DeleteTable requests running simultaneously (in any combination). In other words, the total number of tables in the CREATING, UPDATING or DELETING state cannot exceed 10.
> The only exception is when you are creating a table with one or more secondary indexes. You can have up to 5 such requests running at a time; however, if the table or index specifications are complex, DynamoDB might temporarily reduce the number of concurrent requests below 5.

**Comment by nenti** [comment]
@HyperBrain Thx for pointing that out to me. But this can be handled by serverless framework by queueing table creation during deployment when it exceeds table limit.

**Comment by nenti** [comment]
Quickfix for anyone else experiencing this issue.

1. Delete your dynamodb ressources down to half. 
2. Serverless deploy
3. Restore the full table config
4. Serverless deploy

**Comment by pmuens** [comment]
Thanks for opening @nenti 👍 

> But this can be handled by serverless framework by queueing table creation during deployment when it exceeds table limit.

IMHO this is not the responsibility of the Framework since it's very AWS and DynamoDB specific. The Framework only ensures that the CloudFormation file is compiled and hands it over to AWS.

IMHO the user should ensure that he sticks to the limitations.

**Comment by nenti** [comment]
Ok thank you for pointing that out to me. I create a ticket at AWS. Because if you only put config to AWS then the cloudformation hits the dynamodb limit during automatic provisioning. So this shouldn't be the case.

**Comment by pmuens** [comment]
Thanks for reaching out to AWS @nenti 👍 

> Because if you only put config to AWS then the cloudformation hits the dynamodb limit during 
automatic provisioning. So this shouldn't be the case.

Yes, exactly. It's happening on AWS end.

**Comment by nenti** [comment]
Anyone else hitting the limits of aws dynamodb with cloudformation. AWS support increases the limits quickly if you send support request. No response on resolving this issue in general though.

![image](https://user-images.githubusercontent.com/463984/26919951-bb8d1ce2-4c37-11e7-96b2-b5f0b222ae88.png)

**Comment by mvayngrib** [comment]
there's a good trick mentioned here: https://forums.aws.amazon.com/thread.jspa?threadID=167996

the gist is to add DependsOn to some tables on other tables, to prevent them from being created in parallel

serverless could potentially use this automatically: count the number of tables, number of indexed tables, etc., and inject artificial DependsOn

**Comment by aspiers** [comment]
This is a duplicate of #2622.

**Comment by harleyguru** [comment]
Hi @mvayngrib Do you know how to make dependencies between table and indexes?
For example, let's say I have 5 tables and one table has 2 indexes.
I guess to make the dependencies so that indexes will be created one by one after all base tables has been created.

How could we specify DependsOn property for this?
Thanks

**Comment by mvayngrib** [comment]
@harleyguru I don't think you can as the indexes are not separate CloudFormation resources. Is there a reason you need that specific flow? Maybe it's enough for you to have the tables get created in series, e.g. for tables A-E: E depends on D, D depends on C, and so on. Or E, D depends on C. C, B depends on A.

**Comment by harleyguru** [comment]
First off, Thanks for your response @mvayngrib!
Yeah, My tables have one or 2 indexes in it. Also, I have more than 20 of those tables.
You are right, Indexes can't have DependsOn property.
But DDB doesn't allow simultaneous creation or update of multiple tables with indexes(I tested up to 2 or 3 tables, it worked when total indexes won't exceed 5 across the tables being updated)

Moreover, the important thing is CloudFormation never waits for the activation of indexes being created and starts the updating of other tables(Asynchronously)

This means we or serverless framework should check the completion of indexes creation ourselves.
And upon that status, we should start with next tables dependent on former tables.

But wondering how we should do this via Serverless framework..

**Comment by harleyguru** [comment]
If I've built serverless framework, I should make it to check the response of `describe` operation of table with pending indexes and upon the confirmation of completion CloudFormation should start to tackle with further more tables.
For this, at least it should split the tasks into the multiple steps and check the activation status of indexes in the gap of each steps, and move forward based on that.

Ideally, I wish CF will handle this itself.. or provide any attribute for this so that we don't have to care about the checking of indexes.

**Comment by mvayngrib** [comment]
@harleyguru ah, i wasn't aware of the async index creation. Hm, there must be a workaround... A nasty one that comes to mind is to add a CustomResource lambda per table that will wait for that table's indexes to get created. Then each table can depends on the previous table's corresponding CustomResource lambda

**Comment by harleyguru** [comment]
Yeah, This common task should be handled by anyone certainly..
